src/astar.rs

```diff
@@ -0,0 +1,140 @@
+use std::collections::{
+    HashMap,
+    BinaryHeap,
+};
+use std::collections::hash_map::Entry::{
+    Occupied,
+    Vacant,
+};
+
+use std::hash::Hash;
+
+use scored::MinScored;
+use super::visit::{
+    EdgeRef,
+    GraphBase,
+    IntoEdges,
+    VisitMap,
+    Visitable,
+};
+
+use algo::Measure;
+
+/// [Generic] A* shortest path algorithm.
+///
+/// Compute the length of the shortest path from `start` to `finish`.
+///
+/// The graph should be `Visitable` and implement `IntoEdges`. The function `edge_cost` should
+/// return the cost for a particular edge, which is used to compute path costs. The function
+/// `estimate_cost` should return the estimated cost to the finish for a particular node, also used
+/// to compute path costs. Edge costs must be non-negative.
+///
+/// Returns a Path of subsequent `NodeId` from start to finish, if one was found.
+pub fn astar<G, F, H, K>(graph: G, start: G::NodeId, finish: G::NodeId,
+                         mut edge_cost: F, mut estimate_cost: H)
+    -> Option<Path<G>>
+    where G: IntoEdges + Visitable,
+          G::NodeId: Eq + Hash,
+          F: FnMut(G::EdgeRef) -> K,
+          H: FnMut(G::NodeId) -> K,
+          K: Measure + Copy,
+{
+    let mut visited = graph.visit_map();
+    let mut visit_next = BinaryHeap::new();
+    let mut scores = HashMap::new();
+    let mut path_tracker = PathTracker::<G>::new();
+
+    let zero_score = K::default();
+    scores.insert(start, zero_score);
+    visit_next.push(MinScored(estimate_cost(start), start));
+
+    while let Some(MinScored(_, node)) = visit_next.pop() {
+        if node == finish {
+            return Some(path_tracker.reconstruct_path_to(finish));
+        }
+
+        // Don't visit the same node several times, as the first time it was visited it was using
+        // the shortest available path.
+        if visited.is_visited(&node) {
+            continue
+        }
+
+        visited.visit(node);
```

Comment:
We can combine the is_visited check with this visit call -- using the return value from visit. It returns false if it's already visited.

Reply:
*(re-posting the comment to close the damn Github review)*

Nice catch. I'll edit the comment to something like *"Don't explore the same node's children several times [...]"*, as visiting implies that there's a visitor other than the `VisitMap`.